### PR TITLE
LMDE check added to `check_forked()` function

### PR DESCRIPTION
**Pull request by justsml**
~~Alternate proposal #16498~~

### CODE REVIEW

**Comment by tiborvass** [comment]
@justsml can you extract any useful information from `/etc/os-release` ? If so, I'd prefer using that as a fallback in the case where `lsb_release -u` doesn't work.

**Comment by justsml** [comment]
@tiborvass Is that more like it?

**Comment by fmoliveira** [comment]
@justsml I'm curious, what's inside your `/etc/os-release` file? That looks interesting.

### TL;DR

The use of `/etc/debian_version` is interesting. Trying to go more generic: would it harm Ubuntu distros if we install a Debian package in it? Would the structure differ enough to cause any pain?

If not, we could just check for `debian_version` file and always consider it's a Debian distro in this case, if `lsb_release -u` have failed previously. We'd reach any Debian derivative if this is possible. Some research is needed.

### If you've got the time :sweat_smile: 

The `/etc/debian_version` file is an interesting point. I've got that file in my elementary OS (Ubuntu fork) too and it contains `jessie/sid`. So, checking for its presence may lead to confusion between Debian and Ubuntu distros, that's why I think @justsml have checked his `os-release` file as well.

:question:  **Does the flow below make sense? Would it work?**

   1. *(existing step)* Check `lsb_release -u`, in case of success, consider its upstream info as the script is doing today.
   2. *(new step)* If above fails, check for the presence of `/etc/debian_version` file. If it exists, consider it's a Debian distro and get the corresponding version from the file. Follow the `install.sh` flow of a Debian distro.

:warning:  **Caution:**

The step **2** would be true in Ubuntu forks in which `lsb_release -u` isn't supported. The effect of this would be installing a Debian package in an Ubuntu fork. It doesn't look dangerous at first sight, though I don't know how both distros differs in structure from each other. We would need some research and pick a few example distros to test it.

I know @justsml have suggested using `/etc/os-release` together to check if it's a Linux Mint distro, but I'm trying to go the most generic way we can do. Would installing a Debian package to Ubuntu do any harm? Are there significant structure differences between them?

:bulb:  **What do you think?**

This approach attempts to be more generic so we can reach more distros with less effort. But it may not work.

 I've got some ISOs here, so I'll setup some VMs to test it and report the results here.

**Comment by tiborvass** [comment]
@fmoliveira I'm not sure I understand how you can be more generic than using `/etc/os-release` ?
/cc @tianon

**Comment by fmoliveira** [comment]
@tiborvass The presence of the `/etc/debian_version` file points it's any kind of Debian derivative, and that would lead to more generalisation.

I don't know if it's possible to install the Docker package from Debian repositories in Ubuntu distributions, there may be file structure differences. But if that works, it'd do work across all Debian distros, being much more generic.

Yet we should consider this as risky until we understand file system differences between Debian and Ubuntu.

**Comment by fmoliveira** [comment]
@justsml 

By the way, the `install.sh` already has some checks for the `/etc/debian_version` file in order to install Docker on Debian. Why isn't it working on LMDE? I think the first step is trying to understand why the lines below don't work for LMDE:

Lines 155-157:
```sh
if [ -z "$lsb_dist" ] && [ -r /etc/debian_version ]; then
	lsb_dist='debian'
fi
```

Lines 186-196:
```sh
debian)
	dist_version="$(cat /etc/debian_version | sed 's/\/.*//' | sed 's/\..*//')"
	case "$dist_version" in
		8)
			dist_version="jessie"
		;;
		7)
			dist_version="wheezy"
		;;
	esac
;;
```

@tiborvass If the Debian package also works in Ubuntu distros, the code mentioned above could possibly be used as a fallback to the absence of `lsb_release -u` in Ubuntu derivatives. Sorry for being off topic, this PR is about a pure Debian derivative. I'm going back on topic.

**Comment by justsml** [comment]
I'm still reading the comments... But here's more details of my environment:

```sh
dan@dev:~$ cat /etc/debian_version
8.2

dan@dev:~$ cat /etc/os-release 
PRETTY_NAME="Linux Mint LMDE"
NAME="Linux Mint LMDE"
ID=linuxmint
ANSI_COLOR="1;31"

dan@dev:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	LinuxMint
Description:	LMDE 2 Betsy
Release:	2
Codename:	betsy

dan@dev:~$ lsb_release -u
Usage: lsb_release [options]

lsb_release: error: no such option: -u
dan@dev:~$ lsb_release --help
Usage: lsb_release [options]

Options:
  -h, --help         show this help message and exit
  -v, --version      show LSB modules this system supports
  -i, --id           show distributor ID
  -d, --description  show description of this distribution
  -r, --release      show release number of this distribution
  -c, --codename     show code name of this distribution
  -a, --all          show all of the above information
  -s, --short        show requested information in short format

```

**Comment by justsml** [comment]
`$lsb_dist` ends up being "linuxint"

**Comment by justsml** [comment]
LMDE wants to appear like it's own main distro 'linuxmint' - not 'debian'.

Here's where I was going... 
To catch this specific condition:
I check `/etc/os-release` is `linuxmint` AND we have a `/etc/debian_version` file. 

I know it's getting a little specific putting 'linuxmint' in there - but my patch currently fixes the issue with bare minimal changes.

I tested the non-LMDE **Ubuntu-based** Cinnamon/Mint v17.x - and it does NOT have a `/etc/debian_version` file.

**Comment by justsml** [comment]
Hey @fmoliveira thanks for the suggestions.

I updated my code so there is no mention of linuxmint, just a fallback - see around line 80:

```sh
		else
			if [ -r /etc/debian_version ]; then
				lsb_dist=debian
			fi
		fi
```

Hope that looks better, I think I just needed to :zzz: 

I still need to test both the main/Ubuntu-ish Cinnamon & MATE releases. Their 'Debian Edition' counterparts are working & passing.

**Comment by fmoliveira** [comment]
Hi @justsml ,
Oh, my bad. That's right, forked distros will have their own name in $lsb_dist.
Checking only for `/etc/debian_version` file to pretend it's Debian should work for LMDE and other Debian based distros.

But first I want to confirm if this wouldn't harm Ubuntu forks which don't implement `lsb_release -u`, check the research we've done at #15001. As these distros also have the `/etc/debian_version`, they'd behave like Debian and install packages from Debian sources instead of Ubuntu ones. I don't think it'll be a problem, they shouldn't be very different in structure, but it's good to check it out before deploying this fix.

I'll setup some VMs with Ubuntu forks that I've used in #15001 and post the results here.

**Comment by justsml** [comment]
Thanks for pointing out #15001. I've still gotta work out a rough edge or 2.

I setup tons of test VMs.
And It works in all of them EXCEPT Ubuntu 14.04 LTS. Weird.
I'll have it fixed soon.

Side note: I'm glad I checked out [elemetary OS](https://elementary.io/) - it's beautiful, even the animation on a VM was impressive - it's 'workspaces' UI makes OSX's look like it's from some hack design team.

## Test Results/Progress :+1: 

Pre-test I ran system updates & installed virtualbox guest drivers.

```sh
# The test!
curl https://raw.githubusercontent.com/justsml/docker/LMDEv2/hack/install.sh | bash
```
### Target OSs:
[x] elementary OS
[x] Linux Mint 17.2 Cinnamon & MATE
[x] Linux Mint Debian Edition v2 LMDE2 Cinnamon
[x] Fedora 21, 22 & 23 beta
[ ] Cumulus Linux - `et OSS?`
[x] Kali
[ ] Ubuntu 14.04 **Pending Fix**
[x] Kubuntu
[x] Debian
[ ] openSUSE

**Comment by jfrazelle** [comment]
ah cool!

**Comment by jcbrinfo** [comment]
I confirm that this works with the last version of LMDE Betsy.

**Comment by jfrazelle** [comment]
ping @tianon

**Comment by justsml** [comment]
### Idea: Workaround for Hard-Failures in `install.sh`, w/ a fallback 'Distro pick-list' install mode

> I know confusing users w/ an abundance of choices is usually evil, and the current **inspection+detection method reminds me of what has worked very well for browser devs**
> However, **could it be we are ignoring & overlooking a level of fragility at the edges of the linux universe?** 
> I say 'fragile' because each LSB/distro's `case` block has it's own hardcoded expectations and unique checks. (e.g. debian's `dist_version` check only account for v7 & v8 - even though a Debian v9 (stretch)  apparently is available on the repos). 

> **Side Note** Does this sound feasable: Could a more predictable solution featuring **ZERO** hard-coded LSB/dist names work? Instead using Curl+RegEx extraction on the `repo indexes` to identify & show available choices in a menu/pick-list? ... (Does that make sense? I might need to get some sleep...) 

I'm assuming if you are installing docker at cmd line you probably can pick your closest OS match out of a lineup, hopefully.

#### Anyway here's what I had in mind for a minimal 'Menu/Picklist', (no curl, just hard-coded atm)

```sh
# Example w/ the deb/ubuntu flavors
while true; do
    read -p "Choose the distro most similar to your OS: " n
    case $n in
        [1]* ) lsb_dist=debian; dist_version=jessie;  repo=main; break;;
        [2]* ) lsb_dist=debian; dist_version=stretch; repo=main; break;;
        [3]* ) lsb_dist=debian; dist_version=wheezy;  repo=main; break;;
        [4]* ) lsb_dist=ubuntu; dist_version=precise; repo=main; break;;
        [5]* ) lsb_dist=ubuntu; dist_version=trusty;  repo=main; break;;
        [6]* ) lsb_dist=ubuntu; dist_version=utopic;  repo=main; break;;
        [7]* ) lsb_dist=ubuntu; dist_version=vivid;   repo=main; break;;
        [8]* ) lsb_dist=ubuntu; dist_version=wily;    repo=main; break;;
        [9]* ) exit;;
        * ) echo "Please answer yes or no.";;
    esac
    $sh_c "echo deb https://apt.dockerproject.org/repo ${lsb_dist}-${dist_version} ${repo} > /etc/apt/sources.list.d/docker.list"
done
```

**Comment by vdemeester** [comment]
/ping @justsml @tianon @jfrazelle how are we on that ?

**Comment by tianon** [comment]
This PR LGTM (logic seems fair enough).

As for the future of this script, I think we should probably update it to print out a deprecation warning and have it only detect enough of a platform to be able to link to the relevant installation documentation (and for now, then continue with the install but with a warning that it will eventually no longer perform the install).  Then we can slowly (two major releases) transition it to just be a "detect my platform and link me to the right documentation" script.

**Comment by thaJeztah** [comment]
ping @jfrazelle PTAL

**Comment by jfrazelle** [comment]
LGTM and agreed with what @tianon said above re depreciation warning

**Comment by justsml** [comment]
Thanks for the merge @jfrazelle  !!! 

Ok @tianon - sounds good :+1:

**Comment by chespinoza** [comment]
@jcbrinfo I can't install it using LMDE Betsy : 
wget -qO- https://get.docker.com/ | sh

  Either your platform is not easily detectable, is not supported by this
  installer script (yet - PRs welcome! [hack/install.sh]), or does not yet have
  a package for Docker.  Please visit the following URL for more detailed
  installation instructions:

    https://docs.docker.com/en/latest/installation/
